Clean up fuzzing support

M1cha opened the pull request:
See the commit `arch/posix: make fuzzing easier to use` for an up to date description.

M1cha committed:
arch/posix: put fuzzing kconfigs into submenu

Without fuzzing enabled they don't do anything and should not be
selectable.

Signed-off-by: Michael Zimmermann <sigmaepsilon92@gmail.com>

aescolar commented on boards/posix/native_posix/main.c:
```diff
@@ -120,15 +121,50 @@ int main(int argc, char *argv[])
+
+int _posix_zephyr_main(void)
+{
+	int fuzzer_test_one_input(const uint8_t *data, size_t sz);
```
This would be an expected interface which needs to provided by any app that uses this, but there is no description of it anywhere. This should have been defined in a header and properly commented. (anyhow, this comment is mute given the main review comment)

aescolar commented on boards/posix/native_posix/main.c:
```diff
@@ -120,15 +121,50 @@ int main(int argc, char *argv[])
+
 /**
  * Entry point for fuzzing (when enabled). Works by placing the data
  * into two known symbols, triggering an app-visible interrupt, and
```
The old code had comments (like this) trying to explain the interface. The new code lacks new comments, but leaves the old comments behind which start to be misleading. (anyhow, this comment is mute given the main review comment)

aescolar reviewed (changes requested):
I don't think this change overall is a good idea.
I guess what @andyross had done was one possible way of connecting with the fuzzer. @andyross needed to modify the main loop and as such that change was justified in the very top/main() for this board. But still having those 2 kconfig options at the arch level and the assumption that "the way" to interact with the fuzzer is always thru this fake interrupt and waiting for X ticks, already felt like an stretch.
This PR is going deeper by assuming all the app possibly needs to do is what the fuzzing sample @andyross left did, and adding sample app logic into the very top. Given that the motivation seems to be to reduce a few lines of boilerplate from a test app somebody may need to copy, I don't think it is justified.

What seems ok are the changes of uint8-> size_t and menuconfig.

M1cha commented:
@aescolar I get your point, but what is the solution then?  Having to copy >10 non-whitespace lines per test is hardly "a few" in my opinion. Maybe we can provide a library so we can enable this behavior using a kconfig?

On another note, I don't like the `extern uint8_t *posix_fuzz_buf, posix_fuzz_sz;` from the current implementation since that's a very unsafe thing to do. We should add a header that's included by both the posix part and the app. I wonder if there's something like `-Wmissing-prototypes` for variables so this kind of unsafe code can be caught.

aescolar commented:
> Maybe we can provide a library so we can enable this behavior using a kconfig?

That would be nicer.

M1cha committed:
arch/posix: use size_t for posix_fuzz_sz

That's the type of the original argument and we should use it internally as
well to prevent truncating the value.

Signed-off-by: Michael Zimmermann <sigmaepsilon92@gmail.com>

M1cha committed:
arch/posix: make posix_fuzz_buf const

The original buffer is const as well and we're not allowed to modify it.

Signed-off-by: Michael Zimmermann <sigmaepsilon92@gmail.com>

M1cha commented:
@aescolar I"ve reduced this PR to the improvements to the existing code for now.

Where would I put a library that implements the ISR and runs a callback? I mean this whole thing is still posix specific so `lib/` doesn't feel quite right.

> I guess what @andyross had done was one possible way of connecting with the fuzzer. @andyross needed to modify the main loop and as such that change was justified in the very top/main() for this board. But still having those 2 kconfig options at the arch level and the assumption that "the way" to interact with the fuzzer is always thru this fake interrupt and waiting for X ticks, already felt like an stretch.

In that case, wouldn't it be better if we move andys original implementation out of the posix board and instead simply provided a kconfig to either disable or rename the boards `main` entrypoint so you can override it with your own? That way this would also work with other fuzzers. All of the functions that the current, fuzzer-specific implementation uses seem to be non-static already anyway.

M1cha committed:
arch/posix: move fuzz entry doc to the right place

It should document the entry point function, not the DMA-like variable.

Signed-off-by: Michael Zimmermann <sigmaepsilon92@gmail.com>

aescolar commented on boards/posix/native_posix/main.c:
```diff
@@ -35,6 +35,10 @@
 #include "irq_ctrl.h"
 
+#ifdef CONFIG_ARCH_POSIX_LIBFUZZER
+#include "board_fuzzer.h"
```
This header does not exist.

aescolar commented on samples/subsys/debug/fuzz/src/main.c:
```diff
@@ -18,26 +19,26 @@
  */
 int *global_null_ptr;
-static const uint8_t key[] = { 0x9e, 0x21, 0x0c, 0x18, 0x9d, 0xd1, 0x7d };
+static const uint8_t key[] = {0x9e, 0x21, 0x0c, 0x18, 0x9d, 0xd1, 0x7d};
```
Nit: Normally we avoid doing style fixes in lines we don't need to touch.

aescolar commented on boards/posix/native_posix/main.c:
```diff
@@ -155,7 +166,13 @@ int LLVMFuzzerTestOneInput(const uint8_t *data, size_t sz)
 	posix_exec_for(k_ticks_to_us_ceil64(CONFIG_ARCH_POSIX_FUZZ_TICKS));
 
-	return 0;
+	__ASSERT_NO_MSG(posix_fuzz_result_valid);
```
A Zephyr assert is not appropriate at this level. (The Zephyr kernel is not running here).
If you just want to exit, just call posix_exit(<error_code>), if you also want an error message posix_print_error_and_exit().

aescolar commented on boards/posix/native_posix/main.c:
```diff
@@ -134,6 +135,15 @@ int main(int argc, char *argv[])
 const uint8_t *posix_fuzz_buf;
 size_t posix_fuzz_sz;
 
+static int posix_fuzz_result;
```
I'd much rather go with a `native_` prefix instead of `posix_` both for variables and functions.

M1cha replied:
Should I rename posix_fuzz_buf and posix_fuzz_sz as well?

aescolar replied:
That would be nice. But maybe lets do that then in the follow up PR.

aescolar reviewed (changes requested):
> Where would I put a library that implements the ISR and runs a callback? 

It could be in `boards/posix/common/native_fuzzing/`

> In that case, wouldn't it be better if we move andys original implementation out of the posix board and instead simply provided a kconfig to either disable 

Probably yes.

andyross commented:
Most of this looks like good cleanup to me.

FWIW: it's absolutely true that the current API is sort of a kluge.  There are Big Design Questions about how to match fuzzing (which is fundamentally a functional programming technique: it treats the code under test as a procedure to be executed with varying input, more or less assuming that the results differ only by input) to an RTOS (which is an extremely stateful always-running system with no single point of input).  Making the fuzzing input an interrupt seemed like a reasonably obvious choice, but it's certainly not the only one possible.

So I guess my intuition agrees with Alberto: we shouldn't go too far down the road to making this a locked-down/official API until we're sure we know what we want.  But we should absolutely merge the cleanup patches: const correctness and doc location and the like are never a bad idea.

M1cha commented:
> > In that case, wouldn't it be better if we move andys original implementation out of the posix board and instead simply provided a kconfig to either disable
> 
> Probably yes.

I'll probably do that then. I might also have a use case for fuzzing the nrf52_bsim and a refactoring like that might be a good preparation for that.

M1cha commented:
I've renamed the PR to clarify the reduction of the scope. I'll do another PR that implements the mentioned ideas.

aescolar commented:
I think we have a clear agreement and are mostly ready to merge:
https://github.com/zephyrproject-rtos/zephyr/pull/60180/commits/d94c239224a42910ccc33fbbaf6af076173951db 
https://github.com/zephyrproject-rtos/zephyr/pull/60180/commits/68130e7cb4e54581dc361d643b6f4a218c0b0153
https://github.com/zephyrproject-rtos/zephyr/pull/60180/commits/f352a6f972397255ed77ad1b6af53a0732bb47a4
https://github.com/zephyrproject-rtos/zephyr/pull/60180/commits/1437552d289200ad4af54d5275ddf1b9407a5375

The other 2 commits probably should come in the PR in which you move the logic into a selectable mini library.

andyross reviewed (approved):
Refresh +1
